### PR TITLE
Support reduce expression

### PR DESCRIPTION
**Pull request by jievince**
depends on https://github.com/vesoft-inc/nebula-common/pull/346
doc:https://neo4j.com/docs/cypher-manual/current/functions/list/#functions-reduce

### CODE REVIEW

File: tests/tck/features/expression/reduce.feature [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    Then the result should be, in any order:
+      | r   |
+      | -14 |
+
+  Scenario: use a reduce in GO
+    Given a graph with space named "nba"
+    When executing query:
+      """
+      GO FROM "Tony Parker" OVER like WHERE like.likeness != reduce(totalNum = 5, n IN range(1, 3)
+      | $$.player.age + totalNum + n) YIELD like._dst AS id, $$.player.age AS age, like.likeness AS likeness
```
Comment (yixinglu) [inline tests/tck/features/expression/reduce.feature:30]:
```
GO FROM "Tony Parker" OVER like 
WHERE like.likeness != reduce(totalNum = 5, n IN range(1, 3) | $$.player.age + totalNum + n) 
YIELD like._dst AS id, $$.player.age AS age, like.likeness AS likeness
```

File: tests/tck/features/expression/reduce.feature [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+      | "Tim Duncan"        | 42  | 95       |
+      | "LaMarcus Aldridge" | 33  | 90       |
+
+  Scenario: use a reduce in MATCH
+    Given a graph with space named "nba"
+    When executing query:
+      """
+      MATCH p = (n:player{name:"LeBron James"})<-[:like]-(m)
+      RETURN nodes(p)[0].age AS age1, nodes(p)[1].age AS age2, reduce(totalAge = 100, n IN nodes(p)
+      | totalAge + n.age) as r
```
Comment (yixinglu) [inline tests/tck/features/expression/reduce.feature:44]:
```
MATCH p = (n:player{name:"LeBron James"})<-[:like]-(m)
RETURN
  nodes(p)[0].age AS age1,
  nodes(p)[1].age AS age2, 
  reduce(totalAge = 100, n IN nodes(p) | totalAge + n.age) as r
```

File: tests/tck/features/expression/reduce.feature [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+      | 34   | 34   | 168 |
+      | 34   | 33   | 167 |
+      | 34   | 31   | 165 |
+      | 34   | 29   | 163 |
+      | 34   | 37   | 171 |
+      | 34   | 26   | 160 |
+    When executing query:
+      """
+      MATCH p = (n:player{name:"LeBron James"})-[:like]->(m)
+            RETURN nodes(p)[0].age AS age1, nodes(p)[1].age AS age2, reduce(x = 10, n IN nodes(p) | n.age - x) as x
```
Comment (yixinglu) [inline tests/tck/features/expression/reduce.feature:57]: format

File: tests/tck/features/expression/reduce.feature [hunk 4]
```diff
@@ -0,0 +1,61 @@
+# Copyright (c) 2020 vesoft inc. All rights reserved.
+#
+# This source code is licensed under Apache 2.0 License,
+# attached with Common Clause Condition 1.0, found in the LICENSES directory.
+@jie
```
Comment (yixinglu) [inline tests/tck/features/expression/reduce.feature:5]: cleanup

File: tests/tck/features/expression/reduce.feature (hunk 1 above)
Comment (jievince) [reply tests/tck/features/expression/reduce.feature]: Ok, fixed.

File: src/util/ParserUtil.h [hunk 5]
```diff
@@ -125,6 +125,63 @@ class ParserUtil final {
         }
         pred->setFilter(newFilter);
     }
+
+
+    static void rewriteReduce(QueryContext *qctx,
```
Comment (Shylock-Hg) [inline src/util/ParserUtil.h:130]: What's this purpose?

Comment (jievince) [reply src/util/ParserUtil.h:130]: Rewrite it so we can eval it properly. The logic here is as same as  list comprehension and predicate expression

**Review by Aiee — APPROVED** [review]

**Review by yixinglu — APPROVED** [review]

**Pull request merged**